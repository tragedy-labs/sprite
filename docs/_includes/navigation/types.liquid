{%
  include navigation/list.liquid
  title="Types"
  items=site.types
  icon="icons/ts.svg"
%}
