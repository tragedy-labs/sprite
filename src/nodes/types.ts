export { ArcadeDeleteReturnCount } from './delete/returnCount.js';
export { ArcadeResultSortDirection } from './select/from/direction.js';
export { ArcadeSelectTimeoutStrategy } from './select/from/strategy.js';
