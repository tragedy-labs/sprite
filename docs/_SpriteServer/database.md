---

layout: default

title: database

permalink: /SpriteServer/database.html

---

### _SpriteServer_.database

#### Interface

(**databaseName: *string***)

Returns an `SpriteDatabase` instance for the supplied `databaseName`,
using the authorization details of the `SpriteServer` instance.

#### Example

```ts
const server = new SpriteServer({
  username: 'aUser',
  password: 'aPassword',
  address: 'http://localhost:2480',
});

async function databaseExample() {
  try {
    const database = await server.database('aDatabase');
    // returns an instance of SpriteDatabase
    console.log(database.name);
  } catch (error) {
    // handle errors
    console.error(error);
  }
};

databaseExample();
```

