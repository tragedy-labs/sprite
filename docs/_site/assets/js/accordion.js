/*
 *   This content is licensed according to the W3C Software License at
 *   https://www.w3.org/copyright/software-license-2023/
 *
 *   Simple accordion pattern example
 */

'use strict';

class Accordion {
  constructor(domNode) {
    this.root = domNode;
    this.button = this.root.querySelector('button[aria-expanded]');
    this.icon = this.root.querySelector('svg');
    const controlsId = this.button.getAttribute('aria-controls');
    this.content = document.getElementById(controlsId);

    this.open = this.button.getAttribute('aria-expanded') === 'true';

    // add event listeners
    this.button.addEventListener('click', this.onButtonClick.bind(this));
    this.toggle(false);
  }

  onButtonClick() {
    this.toggle(!this.open);
  }

  toggle(open) {
    // don't do anything if the open state doesn't change
    if (open === this.open) {
      return;
    }

    // update the internal state
    this.open = open;

    // handle DOM updates
    this.button.setAttribute('aria-expanded', `${open}`);
    if (open) {
      this.content.removeAttribute('hidden');
      this.icon.classList.add('open');
    } else {
      this.content.setAttribute('hidden', '');
      this.icon.classList.remove('open');
    }
  }

  // Add public open and close methods for convenience
  open() {
    this.toggle(true);
  }

  close() {
    this.toggle(false);
  }
}

const initializeAccordions = () => {
  // init accordions
  const accordions = document.querySelectorAll('.menuItem');
  accordions.forEach((accordionEl) => {
    new Accordion(accordionEl);
  });
};

document.addEventListener('DOMContentLoaded', initializeAccordions);
