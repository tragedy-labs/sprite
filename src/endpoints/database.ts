export const endpoints = {
  query: '/api/v1/query',
  command: '/api/v1/command',
  beginTransaction: '/api/v1/begin',
  commitTransaction: '/api/v1/commit',
  rollbackTransaction: '/api/v1/rollback'
} as const;
