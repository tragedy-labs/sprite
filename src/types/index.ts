export * from './client.js';
export * from './commands.js';
export * from './database.js';
export * from './edge.js';
export * from './operators.js';
export * from './queries.js';
export * from './server.js';
export * from './transaction.js';
