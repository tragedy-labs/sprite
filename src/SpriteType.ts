import { SpriteDatabase } from './SpriteDatabase.js';
import { TypeNames } from './types/database.js';

/**
 * @description Used to define properties of a type in the schema.
 * @experimental
 */
export class SpriteType<S, N extends TypeNames<S>> {
  private _database: SpriteDatabase;
  private _name: N;
  /** The name of the type */
  get name() {
    return this._name;
  }
  constructor(database: SpriteDatabase, typeName: N) {
    this._name = typeName;
    this._database = database;
  }
}
