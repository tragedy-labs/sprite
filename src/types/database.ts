import { ArcadeSelectTimeoutStrategy } from '../nodes/types.js';
import { ISpriteRestClientConnectionParameters } from './client.js';
import { SpriteRestClient } from '../SpriteRestClient.js';
import { SpriteTransaction } from '../SpriteTransaction.js';
import { ValidSuperTypeKey } from './type.js';

/**
 * The possible values for a parameter of a command or query
 */
export type SpriteAllowedParamValues = string | boolean | number;

/**
 * The possible categories of a record in ArcadeDB,
 * as they appear in the record's `@cat` property.
 */
export type ArcadeRecordCategory = 'd' | 'e' | 'v';

/**
 * The possible record types, as expected to be
 * supplied to a `CREATE TYPE` sql operation.
 */
export type ArcadeRecordType = 'document' | 'edge' | 'vertex';

type RecordMetaBase<C extends ArcadeRecordCategory> = {
  /** The id of the record. */
  '@rid': string;
  /** The category of the record. */
  '@cat': C;
  /** The type of the record. */
  '@type': string;
};

/**
 * The base metadata for record in
 * ArcadeDB.
 */
export type RecordMeta<C extends ArcadeRecordCategory = ArcadeRecordCategory> =
  C extends 'e' ? EdgeRecordMeta : RecordMetaBase<C>;

/**
 * The metadata for an Edge record in
 * ArcadeDB
 */
export interface EdgeRecordMeta extends RecordMetaBase<'e'> {
  /** The `@rid` of the vertex the edge goes to */
  '@in': string;
  /** The `@rid` of the vertex the edge comes from */
  '@out': string;
}

/**
 * Adds ArcadeDB record metadata to the types
 * defined in the schema.
 */
export type AsArcadeEdges<S> = {
  [K in keyof S]: S[K] & EdgeRecordMeta;
};

/**
 * Adds ArcadeDB record metadata to the types
 * defined in the schema.
 */
export type AsArcadeRecords<S, C extends ArcadeRecordCategory> = {
  [K in keyof S]: S[K] & RecordMeta<C>;
};

export type AsArcadeDocuments<S> = AsArcadeRecords<S, 'd'>;
export type AsArcadeVertices<S> = AsArcadeRecords<S, 'v'>;

/**
 * The TypeNames in a supplied schema.
 */
export type TypeNames<Schema> = keyof Schema;

/**
 * The parameters necessary to perform opertations
 * on a target ArcadeDB database.
 */
export interface ISpriteDatabaseConnectionParameters
  extends ISpriteRestClientConnectionParameters {
  /**
   * The name of the database to connect to.
   */
  databaseName: string;
}

/**
 * The parameters used to create a `SpriteDatabase`
 * intance when there is an existing instance of a
 * `SpriteRestClient`, such as when intantiated
 * from a method in the `SpriteServer` class
 */
export interface ISpriteDatabaseClientParameters {
  /**
   * The `SpriteRestClient` instance
   */
  client: SpriteRestClient;
  /**
   * The name of the database to connect to.
   */
  databaseName: string;
}

/**
 * The Query languages supported by ArcadeDB, supplied as a parameter
 * to `SpriteDatabase.query()` and `SpriteDatabase.command()`
 */
export type ArcadeSupportedQueryLanguages =
  | 'sql'
  | 'sqlscript'
  | 'graphql'
  | 'cypher'
  | 'gremlin'
  | 'mongo';

/**
 * Options to create a new type with.
 */
export interface ISpriteCreateTypeOptions<S, N extends TypeNames<S>> {
  /**
   * When set to true, the type creation will be ignored if the
   * type already exists (instead of failing with an error).
   * @default false
   */
  ifNotExists?: boolean;
  /**
   * Defines a super-type you want to extend with this type.
   */
  extends?: ValidSuperTypeKey<S, N>;
  /**
   * A bucket-name, or an array of bucket-names you want this type to use.
   */
  buckets?: string | Array<string>;
  /**
   * Defines the total number of buckets you want to create for this type. The
   * @default 1
   */
  totalBuckets?: number;
}

export type OmitMeta<T> = Omit<T, keyof RecordMeta | keyof EdgeRecordMeta>;

/** Options to insert a new record with */
export interface ISpriteInsertRecordOptions<T> {
  /** The data to populate the newly created reacord with */
  data?: OmitMeta<T> | OmitMeta<T>[];
  /** The bucket to store the record in */
  bucket?: string;
}

/**
 * Options for a database.dropType() command
 */
export interface ISpriteDropTypeOptions {
  /**
   * Prevent errors if the type does not exits when attempting to drop it.
   * @default false;
   */
  ifExists?: boolean;
  /**
   * Defines whether the command drops non-empty edge and vertex types. Note, this can
   * disrupt data consistency. Be sure to create a backup before running it.
   * @default: false
   */
  unsafe?: boolean;
}

/** A callback which contains the operations to be conducted within a transaction */
export type SpriteTransactionCallback =
  | ((trx: SpriteTransaction) => Promise<void>)
  | ((trx: SpriteTransaction) => void);

/**
 * How the bucket will be selected for a newly created record of this type
 * @default 'round-robin'
 */
export type ArcadeBucketSelectionStrategies =
  | 'round-robin'
  | 'thread'
  | 'partitioned<primary-key>';

/**
 * A type definition returned by ArcadeDB when a getSchema command
 * is compeleted.
 */
export type ArcadeTypeDefinition = {
  /** The name of the type */
  name: string;
  /** The category of the type (document, vertex, or edge) */
  type: ArcadeRecordType;
  /**
   * The number of records with this type name
   */
  records: number;
  /**
   * The buckets associated with this type.
   */
  buckets: Array<string>;
  /** How the bucket will be selected for a newly created record of this type */
  bucketSelectionStrategy: ArcadeBucketSelectionStrategies;
  /**
   * Super types associated with this type.
   */
  parentTypes: Array<string>;
  /**
   * Properties defined for this type.
   */
  properties: Array<string>;
  /** Indexes defined for records of this type */
  indexes: Array<string>;
  /** An object for custom user-defined properties for this type */
  custom: Record<string, unknown>;
};

/** The response from a `SpriteDatabase.getSchema()` query */
export type ArcadeGetSchemaResponse = ArcadeTypeDefinition[];

export type ArcadeSqlExaplantionExecutionPlanStep = {
  cost: number;
  description: string;
  javaType: string;
  name: string;
  subSteps: ArcadeSqlExaplantionExecutionPlanStep[];
  targetNode: string;
  type: string;
};

/** SQL Explanation Execution Plan */
export type ArcadeSqlExplanationExecutionPlan = {
  type: 'QueryExecutionPlan' | 'DDLExecutionPlan';
  javaType: string;
  cost: number;
  prettyPrint: string;
  steps: ArcadeSqlExaplantionExecutionPlanStep[];
};

/**
 * An object representing the explanation of an
 * SQL command.
 */
export type ArcadeSqlExplanation = {
  executionPlan: ArcadeSqlExplanationExecutionPlan;
  executionPlanAsString: string;
};

/**
 * The object returned from the server following a successfully command
 */
export interface ArcadeCommandResponse<T = unknown> {
  /**
   * The user that issued the command / query.
   */
  user: string;
  /** The version of ArcadeDB returning the result. */
  version: string;
  /**
   * The name of the server returning the result.
   */
  serverName: string;
  /**
   * The result of the command / query.
   */
  result: T;
}

/**
 * The object returned from the server following a successfully query
 */
export type ArcadeQueryResponse<T = unknown> = ArcadeCommandResponse<T>;

// TODO: have not checked these for compatibility with ArcadeDB
export declare const COMPARISON_OPERATORS: readonly [
  '=',
  '==',
  '!=',
  '<>',
  '>',
  '>=',
  '<',
  '<=',
  'in',
  'not in',
  'is',
  'is not',
  'like',
  'not like',
  'match',
  'ilike',
  'not ilike',
  '@>',
  '<@',
  '&&',
  '?',
  '?&',
  '!<',
  '!>',
  '<=>',
  '!~',
  '~',
  '~*',
  '!~*',
  '@@',
  '@@@',
  '!!',
  '<->',
  'regexp',
  'is distinct from',
  'is not distinct from'
];

/** Operators for a `WHERE` sql statement */
export type SpriteOperators = (typeof COMPARISON_OPERATORS)[number];

/** Options for a `selectFrom` operation, as executed via a `SpriteDatabase` modality. */
export interface ISpriteSelectFromOptions<
  S,
  N extends keyof S,
  P extends keyof WithRid<S, N>
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where?: SpriteWhereClause<S, N, P>;
  /**
   * Designates the field with which to order the result-set.
   * Use the optional 'ASC' and 'DESC' operators to define the direction of the order.
   */
  orderBy?: {
    field: keyof S[N];
    /**
     * Defines the direction to sort the result (ASCending or DESCending).
     * @default 'ASC'.
     */
    direction?: 'ASC' | 'DESC';
  };
  /**
   * Defines the number of records you want to skip from the start of the result-set.
   * You mayfind this useful in Pagination, when using it in conjunction with the
   * limit `option`.
   */
  skip?: number;
  /**
   * Defines the maximum number of records in the result-set. You may find this useful in
   * Pagination, when using it in conjunction with the `skip` option.
   */
  limit?: number;
  /**
   * Defines the maximum time in milliseconds for the query, and optionally the
   * exception strategy to use.
   */
  timeout?: {
    /**
     *  The duration of the timeout in milliseconds.
     */
    duration: number;
    /**
     * The timeout strategy to use.\
     * `RETURN` Truncates the result-set, returning the data collected up to the timeout.\
     * `EXCEPTION` Raises an exception.
     */
    strategy?: ArcadeSelectTimeoutStrategy;
  };
}

/**
 * A base type, with an `@rid` property added.
 */
export type WithRid<S, N extends TypeNames<S>> = OmitMeta<S[N]> & {
  '@rid': string;
};

/**
 * An array with three items used to describe a `where` statement.
 * @example ['@rid', '==', '#0:0']
 */
export type SpriteWhereClause<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>
> = [P, SpriteOperators, WithRid<S, N>[P]];

/**
 * Options for a `deleteFrom` operation as executed
 * via a `SpriteDatabase` modality.
 */
export interface ISpriteDeleteFromOptions<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where: SpriteWhereClause<S, N, P>;
  /**
   * The duration of the timeout in milliseconds.
   */
  timeout?: number;
  /**
   * Defines the maximum number of records in the result-set.
   * @default undefined
   */
  limit?: number;
  /**
   * Defines what is returned following the command: the count of the records before (`BEFORE`) or following deletion (`COUNT`).
   * @default 'COUNT'
   */
  return?: 'COUNT' | 'BEFORE';
}
