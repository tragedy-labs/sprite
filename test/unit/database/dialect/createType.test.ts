import { client } from './testClient.js';
import { endpoints } from '@/endpoints/database.js';
import { variables, headers } from '../../../variables.js';
import { SpriteType } from '@/SpriteType.js';

import { DocumentTypes } from '../../types.js';

const recordType = 'document';
const typeName = 'aDocument';
const SpriteDatabase = client.database;
type TypeName = typeof typeName;

const createTypeResult = {
  user: variables.username,
  serverName: '',
  version: '',
  result: [{ typeName }]
};
describe('SqlDialect.createType()', () => {
  it(`makes a properly formatted POST request to ${endpoints.command}/${variables.databaseName}`, async () => {
    // Arrange
    jest.spyOn(global, 'fetch').mockResolvedValueOnce({
      status: 200,
      json: async () => createTypeResult
    } as Response);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType);

    // Assert
    expect(global.fetch).toHaveBeenCalledWith(
      `${variables.address}${endpoints.command}/${variables.databaseName}`,
      {
        method: 'POST',
        headers,
        body: JSON.stringify({
          language: 'sql',
          command: `CREATE document TYPE ${typeName}`
        })
      }
    );
  });

  it('returns an instance of SpriteType when it receives a JSON object with a result property from the server', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce(createTypeResult.result);

    // Act
    const result = await client.createType<DocumentTypes, TypeName>(
      typeName,
      recordType,
      { buckets: variables.bucketName }
    );

    // Assert
    expect(result).toBeInstanceOf(SpriteType);
  });

  it('handles "buckets" option by appending "BUCKET" + [bucketName] to the command when passed a string bucketName property', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce(createTypeResult.result);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType, {
      buckets: variables.bucketName
    });

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE document TYPE ${typeName} BUCKET ${variables.bucketName}`
    );
  });

  it('handles "buckets" option by appending "BUCKET" + [bucketNames] to the command when passed an array of strings as the bucketName property', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce(createTypeResult.result);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType, {
      buckets: [variables.bucketName, variables.bucketName]
    });

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE document TYPE ${typeName} BUCKET ${[variables.bucketName, variables.bucketName].join(',')}`
    );
  });

  it('handles "ifNotExists" option by appending "IF NOT EXISTS" to the command when passed true', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockImplementationOnce(async () => createTypeResult.result);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType, {
      ifNotExists: true
    });

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE document TYPE ${typeName} IF NOT EXISTS`
    );
  });

  it('handles "extends" option by appending "EXTENDS" + [typeName] to the command when passed an extends property', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce(createTypeResult.result);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType, {
      extends: 'anotherDocument'
    });

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE document TYPE ${typeName} EXTENDS anotherDocument`
    );
  });

  it('handles "totalBuckets" option by appending "BUCKETS" + [totalBuckets] to the command when passed a totalBuckets property', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce(createTypeResult.result);

    // Act
    await client.createType<DocumentTypes, TypeName>(typeName, recordType, {
      totalBuckets: 4
    });

    // Assert
    expect(SpriteDatabase.command).toHaveBeenCalledWith(
      'sql',
      `CREATE document TYPE ${typeName} BUCKETS 4`
    );
  });

  it('should throw an error if it receives an empty string for parameters', async () => {
    // @ts-expect-error - Testing for empty string
    await expect(client.createType('')).rejects.toMatchSnapshot();
  });

  it('should throw an error if it receives no arguments', async () => {
    // @ts-expect-error - Testing for no arguments
    await expect(client.createType()).rejects.toMatchSnapshot();
  });

  it('should throw an error if the server response result is an empty array, and ifNotExists is not true', async () => {
    // Arrange
    jest.spyOn(SpriteDatabase, 'command').mockResolvedValueOnce([]);

    // Act
    await expect(
      client.createType<DocumentTypes, TypeName>(typeName, recordType)
    ).rejects.toMatchSnapshot();
  });

  it('should throw an error if the server response result contains an array containing an object that does not contain the created typename', async () => {
    // Arrange
    jest
      .spyOn(SpriteDatabase, 'command')
      .mockResolvedValueOnce([{ typeName: 'EXPECT_AN_ERROR' }]);

    // Act
    await expect(
      client.createType<DocumentTypes, TypeName>(typeName, recordType)
    ).rejects.toMatchSnapshot();
  });

  it('should return an instance of SpriteType if the server responds appropriately', async () => {
    // Arrange
    jest.spyOn(SpriteDatabase, 'command').mockResolvedValueOnce([{ typeName }]);

    // Act
    const result = await client.createType<DocumentTypes, TypeName>(
      typeName,
      recordType
    );

    // Assert
    expect(result).toBeInstanceOf(SpriteType);
  });

  it('should return an instance of SpriteType if response is an empty array and ifNotExists is true', async () => {
    // Arrange
    jest.spyOn(SpriteDatabase, 'command').mockResolvedValueOnce([]);

    // Act
    const result = await client.createType<DocumentTypes, TypeName>(
      typeName,
      recordType,
      { ifNotExists: true }
    );

    // Assert
    expect(result).toBeInstanceOf(SpriteType);
  });

});
