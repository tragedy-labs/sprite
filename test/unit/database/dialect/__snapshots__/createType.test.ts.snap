// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlDialect.createType() should throw an error if it receives an empty string for parameters 1`] = `[Error: Unable to create type: [''], in database: ['Test']]`;

exports[`SqlDialect.createType() should throw an error if it receives no arguments 1`] = `[Error: Unable to create type: ['undefined'], in database: ['Test']]`;

exports[`SqlDialect.createType() should throw an error if the server response result contains an array containing an object that does not contain the created typename 1`] = `[Error: Unable to create type: ['aDocument'], in database: ['Test']]`;

exports[`SqlDialect.createType() should throw an error if the server response result is an empty array, and ifNotExists is not true 1`] = `[Error: Unable to create type: ['aDocument'], in database: ['Test']]`;
